Fix rabbitmq_policy tags json.loads

nazunalika opened the pull request:
##### SUMMARY
Fixes #28 by removing the json.loads piece of `tags =` in `_policy_check`

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
`rabbitmq_policy`

##### ADDITIONAL INFORMATION
During the `_policy_check` piece of this module, `policy_data` is basically a list based on a split of the variable `policy`. However, `policy` does not contain data. As a result of the split, it becomes: 

`['vhost', 'name', 'pattern', 'apply-to', 'definition', 'priority']`

As there is no JSON data, there is no reason to use `json.loads` on a string.

<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
To reproduce this issue:
1. Install the ansible collection using ansible-galaxy
2. Create a playbook with a set of tasks that deal with rabbitmq policies on a vhost. Example tasks below.

```
- name: Configure pubsub virtual host
  community.rabbitmq.rabbitmq_vhost:
    name: /pubsub
    state: present

- name: Configure an HA policy for pubsub
  community.rabbitmq.rabbitmq_policy:
    name: HA
    apply_to: queues
    pattern: ".*"
    tags:
      ha-mode: all
      ha-sync-mode: automatic
      ha-sync-batch-size: 10000
    vhost: /pubsub
```

3. Run the playbook

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
# BEFORE
The full traceback is:
Traceback (most recent call last):
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 102, in <module>
    _ansiballz_main()
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.rabbitmq.plugins.modules.rabbitmq_policy', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 248, in <module>
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 237, in main
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 161, in has_modifications
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 161, in <genexpr>
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 198, in _policy_check
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

#AFTER
TASK [rockylinux.rabbitmq : Configure the HA policy for pubsub] *********************************************
changed: [rabbitmq001.example.com]
```

odyssey4me commented:
Could you add an integration test to validate that this works so that it continues to work for any further work done to the module? Feel free to also add more integration tests for this module - it unfortunately does not have any at this time.

odyssey4me reviewed (changes requested):
Could you add an integration test to validate that this works so that it continues to work for any further work done to the module? Feel free to also add more integration tests for this module - it unfortunately does not have any at this time.

nazunalika commented:
I added some integration tests but I'm unsure of why they're failing. I'm not sure if I did them incorrectly, so I'm open to review.

Andersson007 commented:
Hi @nazunalika , thanks for the PR! I close-reopened it to trigger the tests. There are attempts to fix the CI, I hope they will succeed. Thanks for your work!

Andersson007 commented:
could you please rebase the PR https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html#rebasing-your-branch

Andersson007 commented:
There are assertion failings like below
```
02:16 TASK [rabbitmq_policy : Check that the host is added] **************************
02:16 task path: /root/ansible_collections/community/rabbitmq/tests/output/.tmp/integration/rabbitmq_policy-0b985dmn-ÅÑŚÌβŁÈ/tests/integration/targets/rabbitmq_policy/tasks/tests.yml:16
02:16 fatal: [testhost]: FAILED! => {
02:16     "assertion": "result is changed",
02:16     "changed": false,
02:16     "evaluated_to": false,
02:16     "msg": "Assertion failed"
02:16 }
```

Andersson007 commented:
it could mean:
1) the change broke something
2) same was before you changed the code

In this case, if i were you, i would 1) revert the change you made in the module 2) commit and push 3) depending on if the tests pass of don't, will fix the code correspondingly

Andersson007 commented:
If the tests pass without you changes in the module, we could merge them here and you could open another PR to fix the issue you were going to fix (because the tests themselves are a very valuable addition).

nazunalika commented:
I am not sure if there is something wrong with the test system here or otherwise. I copied the vhost pieces from the integration tests to help supplement the policy tests. I also tried to rename the vhost from `/test` to `/policytest` - I figured that wouldn't help, but I gave it a try anyway.

I am able to run the tests by hand with success with my one line change to `plugins/modules/rabbitmq_policy.py` from this PR.

Below is my playbook.

```
---
- hosts: rabbit.angelsofclockwork.net
  become: yes
  gather_facts: true

  tasks:
    - name: Set the correct facts
      set_fact:
        vhost_name: /policytest
        ha_policy_name: HA

    - name: Add host
      rabbitmq_vhost:
        name: "{{ vhost_name }}"
        state: present
      register: result

    - name: Chcek that the host was created successfully
      shell: "rabbitmqctl list_vhosts name tracing | grep /policytest"
      register: ctl_result

    - name: Check that the host is added
      assert:
        that:
          - result is changed
          - result is success
          - '"false" in ctl_result.stdout'

    - name: Add host (idempotency)
      rabbitmq_vhost:
        name: "{{ vhost_name }}"
        state: present
      register: result
  
    - name: Check idempotency
      assert:
        that:
          - result is not changed
  
    - name: Add an HA Policy
      rabbitmq_policy:
        name: "{{ ha_policy_name }}"
        apply_to: queues
        pattern: ".*"
        tags:
          ha-mode: all
          ha-sync-mode: automatic
          ha-sync-batch-size: 10000
        vhost: "{{ vhost_name }}"
      register: add_policy
  
    - name: Check that the policy is added
      assert:
        that:
          - add_policy is changed
          - add_policy is success
  
    - name: Add an HA Policy (idempotency)
      rabbitmq_policy:
        name: "{{ ha_policy_name }}"
        apply_to: queues
        pattern: ".*"
        tags:
          ha-mode: all
          ha-sync-mode: automatic
          ha-sync-batch-size: 10000
        vhost: "{{ vhost_name }}"
      register: add_policy
  
    - name: Check policy idempotency
      assert:
        that:
          - add_policy is not changed
  
    - name: Remove the HA policy
      rabbitmq_policy:
        name: "{{ ha_policy_name }}"
        state: absent
        vhost: "{{ vhost_name }}"
      register: remove_policy
  
    - name: Check that the policy is removed
      assert:
        that:
          - remove_policy is changed
          - remove_policy is success
  
    - name: Remove the HA Policy (idempotency)
      rabbitmq_policy:
        name: "{{ ha_policy_name }}"
        state: absent
        vhost: "{{ vhost_name }}"
      register: remove_policy
  
    - name: Check that the policy is removed (idempotency)
      assert:
        that:
          - remove_policy is not changed
```

The result:

```
PLAY [rabbit.angelsofclockwork.net] *********************************************************************************

TASK [Gathering Facts] **********************************************************************************************
ok: [rabbit.angelsofclockwork.net]

TASK [set some facts] ***********************************************************************************************
ok: [rabbit.angelsofclockwork.net]

TASK [Add host] *****************************************************************************************************
changed: [rabbit.angelsofclockwork.net]

TASK [Chcek that the host was created successfully] *****************************************************************
changed: [rabbit.angelsofclockwork.net]

TASK [Check that the host is added] *********************************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Add host (idempotency)] ***************************************************************************************
ok: [rabbit.angelsofclockwork.net]

TASK [Check idempotency] ********************************************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Add an HA Policy] *********************************************************************************************
changed: [rabbit.angelsofclockwork.net]

TASK [Check that the policy is added] *******************************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Add an HA Policy (idempotency)] *******************************************************************************
ok: [rabbit.angelsofclockwork.net]

TASK [Check policy idempotency] *************************************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Remove the HA policy] *****************************************************************************************
changed: [rabbit.angelsofclockwork.net]

TASK [Check that the policy is removed] *****************************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Remove the HA Policy (idempotency)] ***************************************************************************
ok: [rabbit.angelsofclockwork.net]

TASK [Check that the policy is removed (idempotency)] ***************************************************************
ok: [rabbit.angelsofclockwork.net] => {
    "changed": false,
    "msg": "All assertions passed"
}

PLAY RECAP **********************************************************************************************************
rabbit.angelsofclockwork.net  : ok=15   changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Andersson007 commented:
@nazunalika thanks for the feedback!
I ran the tests locally from you fork with `ansible-test integration rabbitmq_policy --docker ubuntu2004 -vvv > ~/test.log`
1) it failed as well with your changes
2) and it passes after i returned `json.loads()`

Also:
3) I'm not sure if the PR fixes the issue from the description. Is it about README, isn't it?

I guess the issue can be related to server version. I know how we can fix it.
Give me a second, I'll change the code.

Andersson007 commented on plugins/modules/rabbitmq_policy.py:
```diff
@@ -195,7 +195,7 @@ def _policy_check(self,
         apply_to = policy_data[apply_to_fno]
         pattern = policy_data[pattern_fno].replace('\\\\', '\\')
-        tags = json.loads(policy_data[tags_fno])
+        tags = policy_data[tags_fno]
```
```suggestion
        try:
            tags = json.loads(policy_data[tags_fno])
        except JSONDecodeError:
            tags = policy_data[tags_fno]
```
Continuing https://github.com/ansible-collections/community.rabbitmq/pull/58#issuecomment-884683648.

Could you please apply the changes and try to run your test playbook? It should work

Also I can see that you made the changes in `main`. I would suggest using a separate branch, see the [Quick-start guide](https://github.com/ansible/community-docs/blob/main/create_pr_quick_start_guide.rst). Otherwise, if we merge the PR now, you'll probably can't continue your contribution with your fork and you'll have to remove your fork and re-fork it again.

I'm looking forward for your feedback, thank you!

Andersson007 replied:
If everything is fine after applying what i'm suggesting and local testing, could you also please add a changelog fragment https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment, so users will know about the bugfix.

nazunalika commented:
There doesn't appear to be a way for me to switch which source branch I'm using for this PR. I plan on submitting a new PR soon instead, hopefully following your suggestions correctly. I will have it tagged as `[WIP]` first. Thank you so much

Andersson007 commented:
Thanks! I left the comment in the new one.
